src/ConnectionPool.jl

```diff
@@ -459,14 +476,49 @@ function keepalive!(tcp)
     return
 end
 
+struct ConnectTimeout <: Exception
+    host
+    port
+end
+
 function getconnection(::Type{Sockets.TCP},
                        host::AbstractString,
                        port::AbstractString;
                        keepalive::Bool=false,
+                       connect_timeout::Int=0,
                        kw...)::Sockets.TCP
+
     p::UInt = isempty(port) ? UInt(80) : parse(UInt, port)
+
     @debug 2 "TCP connect: $host:$p..."
-    tcp = Sockets.connect(Sockets.getaddrinfo(host), p)
+
+    if connect_timeout == 0
+        tcp = Sockets.connect(Sockets.getaddrinfo(host), p)
+        keepalive && keepalive!(tcp)
+        return tcp
+    end
+
+    tcp = Sockets.TCPSocket()
+    Base.connect!(tcp, Sockets.getaddrinfo(host), p)
+
+    timeout = Ref{Bool}(false)
+    @schedule begin
+        sleep(connect_timeout)
+        if tcp.status == Base.StatusConnecting
+            timeout[] = true
+            ccall(:jl_forceclose_uv, Void, (Ptr{Void},), tcp.handle)
+            #close(tcp)
+        end
+    end
+    try
+        Base.wait_connected(tcp)
+    catch e
+        if timeout[]
+            throw(ConnectTimeout(host, port))
```

Comment:
Why do you want to make sure it throws this exception?

Reply:
Because I don't think `UVError` should ever be thrown from a user-level API unless it means there is a bug. See https://github.com/JuliaLang/julia/pull/15879 etc.

I think application level code will want to handle "I set a short timeout and it wasn't met" differently from "something bad happened deep in the network stack.

Elsewhere in HTTP.jl we wrap things like `UVError` and `ArgumentError "stream is closed or unusable"` in an `IOError` type:
https://github.com/JuliaWeb/HTTP.jl/blob/master/src/ConnectionRequest.jl#L36

In higher level code it is important to handle some specific errors by retrying but make sure everything else finds its way up to create a crash and a stack trace and an issue report and a fix. e.g. https://github.com/JuliaCloud/AWSCore.jl/blob/master/src/http.jl#L47-L51

Reply:
`ArgumentError "stream is closed or unusable"` seems pretty wrong because it looks like a precondition error, but there is now way to check the precondition before calling the API because the stream be `isopen()` before the call but be found to be "closed or unusable" while the call is running.

Reply:
Hi @vtjnash, I just noticed that `poll_fd` has a `timeout_s` option. Would that work here?

Reply:
I'm not sure how that's relevant here, since you don't have a `fd`. But yes, that does show another pattern that can be used to implement timeouts (with some more difficulty ... in fact, you've just inspired me to rewrite that file better ... you may see a new PR to base tomorrow sometime 🙂)

Reply:
I was assuming that I could get the `fd` out of the socket with `uv_fileno` or something like that.
Is there a gotcha along the lines of windows socket handles won't work with `poll_fd`?

> new PR

👍🙃

Reply:
(a) the socket isn't open, so you can't poll it (b) it's invalid to pass libuv-derived fd handles to libuv

Reply:
thx for taking the time to explain :)